### ISSUE TITLE
Alignment of Client Credential authentication / authorization using API Gateway Pattern

### ISSUE THREAD

**Issue by mhfoo** [camaraproject/IdentityAndConsentManagement#104]
**Problem description**
Referring to the following:
[CIBA Backend Flow](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-API-access-and-user-consent.md#ciba-flow-backend-flow)

```
POST /bc-authorize HTTP/1.1
Authorization: Basic {Credentials}
Content-Type: application/x-www-form-urlencoded

login_hint=tel%3A%2B346xxyyyzzz&
… 
```

The authorization should follow [API Gateway Pattern](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-AuthN-AuthZ-Concept.md#api-gateway-pattern-) with [Client Credential Grant](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-AuthN-AuthZ-Concept.md#client-credential-grant) access token usage, instead of Basic authentication.

**Expected behavior**
Align with API Gateway Pattern mentioned above.

**Alternative solution**
N.A.

**Additional context**
GSMA has a concept of a north star where the API consumer authentication / authorization is the same implementation for all Operator Open Gateway CIBA implementation and only a change in credentials is sufficient for a change in API provider.

**Comment by AxelNennker** [comment]
We could follow the CIBA example and remove the Authorization header from examples.
That is what I suggest.

CIBA already states:
"Because in the CIBA flow, the OP does not have an interaction with the end-user through the consumption device, it is REQUIRED that the Client provides one (and only one) of the hints specified above in the authentication request, that is "login_hint_token", "id_token_hint" or "login_hint". "

The examples in CIBA do not mention the "Authorization" header.

But CIBA already states that the request must be authenticated.
"The Client MUST authenticate to the Backchannel Authentication Endpoint using the authentication method registered for its client_id, such as the authentication methods from Section 9 of [[OpenID.Core]](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#OpenID.Core) or authentication methods defined by extension in other specifications"

So, even when the Authorization header is not present in examples it is clear that the authorization request must be authenticated. It could also be a signed request.
https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#signed_auth_request

**Comment by eric-murray** [comment]
The CIBA `bc-authorize` example is just that - an example. I assume `client_secret_basic` authentication was used as that is the default, but it is not mandated. Discussions on OpenID Connect profiles for CAMARA have not really started, but I'd expect `private_key_jwt` to be the preferred option. This is the example given in the CIBA specification, and is also used in the GSMA OGW reference implementation.

Requesting a bearer token to allow you to submit a subsequent request for another bearer token seems unnecessarily convoluted, and will not help latency at all.

**Comment by mhfoo** [comment]
@jpengar mentioned the following client credentials in the 12 Jan, Open Gateway Weekly Technical Drop In Sessions (GitHub) meeting.

https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication

**Comment by jpengar** [comment]
Just to avoid naming misunderstandings. The link I shared is what the OpenID Connect core specification defines for **Client Authentication**. Consider the client role definition per [OAuth 2.0 - RFC6749](https://www.rfc-editor.org/rfc/rfc6749.html#section-1.1). The client is generally an application that makes protected resource requests on behalf of and with the permission of the resource owner.
And usually when we talk about "Client Credentials" we are referring to [Client Credentials authorization grant](https://www.rfc-editor.org/rfc/rfc6749.html#section-1.3.4).

**Comment by mhfoo** [comment]
@jpengar @eric-murray @shilpa-padgaonkar 

The following diagram is extracted from [CAMARA-AuthN-AuthZ-Concept.md](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-AuthN-AuthZ-Concept.md#client-credential-grant) under the "Client Credential Grant" section.

![cc](https://github.com/camaraproject/IdentityAndConsentManagement/assets/134043190/7a797bd3-e6c5-4eeb-a2eb-027daafad0c4)

In the above diagram, the "Authorization Server" and "Resource Service" aka CAMARA Service are protected by the API Gateway. The diagram illustrates the "Token request" and "Token Usage" flow and the involved components.

**Question:**
Should the CIBA endpoint "POST /bc-authorize" be protected following the above "Token request" and "Token Usage" flows and API Gateway?

Suggest updating the documentation to mention the intended Client Credential flows and remove items which are not applicable, to achieve a common behaviour.

**Comment by eric-murray** [comment]
@mhfoo

Valid client authentication methods for the CIBA `/bc-authorize` endpoint are discussed [here](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#rfc.section.7.1):

_The Client MUST authenticate to the Backchannel Authentication Endpoint using the authentication method registered for its client_id, such as the authentication methods from Section 9 of [[OpenID.Core]](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#OpenID.Core) or authentication methods defined by extension in other specifications._

So basically, anything goes. **CAMARA have not yet specified which method(s) are allowed**, but the [GSMA OGW reference implementation](https://github.com/GSMA-Open-Gateway/opengateway-reference-implementation) uses `private_key_jwt`.

**Comment by mhfoo** [comment]
#103 will delete CAMARA-AuthN-AuthZ-Concept.md

**Comment by mhfoo** [comment]
> The main objection to the example seemed to be the use of "Basic". I suggest to either replace "Basic" by some other method or leave the Authorization header value blank, indicating that the heaser is needed but the example does not provide a value.

@AxelNennker
Yes, using "Basic" is a security concern.

Understand the community will be aligned to using FAPI 2.0 Security Profile, ```private_key_jwt``` with ```jwks_uri```
Telefónica will contribute their version as a starting point.

**Comment by mhfoo** [comment]
[IdentityAndConsentManagement PR 113](https://github.com/camaraproject/IdentityAndConsentManagement/pull/113) 

Will answer this question

**Comment by jpengar** [comment]
> [IdentityAndConsentManagement PR 113](https://github.com/camaraproject/IdentityAndConsentManagement/pull/113)
> 
> Will answer this question

@mhfoo Then, if it's okay with you, I suggest to close this issue. Just make sure you review PR #113 to verify that the OIDC profile being defined is aligned with Signtel's expectations and needs.

**Comment by jpengar** [comment]
> The main objection to the example seemed to be the use of "Basic". I suggest to either replace "Basic" by some other method or leave the Authorization header value blank, indicating that the header is needed but the example does not provide a value.

@AxelNennker @mhfoo We can complete and merge PR #105 to fix this issue.

**Comment by nitroduna** [comment]
I don't know if this is already resolved, but just to clarify. The bc-authorize endpoint has to validate the client_id (APP) credentials. As the spec says this have to be done using the same methods used in another OAuth2 endpoints that require client_id (APP) credentials, such as the token endpoint. 

So any of these methods are available:
- Authorization header with Basic auth including the "client_id:client_secret" encoded in Base64
- Using private_key_jwt method, which is done with the parameter "client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer" and the parameter "client_assertion" with the signed JWT assertion.

Both systems are valid to any protected OAuth2 endpoint and therefore valid for bc-authorize endpoint. 

I think that in CAMARA, the "private_key_jwt" authentication is the method selected for any auth endpoint that requires APP authentication or credentials. If that so, the bc-authorize should use the same authentication method.

**Comment by mhfoo** [comment]
@nitroduna yes, 'private_key_JWT' is the preferred method.

**Comment by mhfoo** [comment]
> > [IdentityAndConsentManagement PR 113](https://github.com/camaraproject/IdentityAndConsentManagement/pull/113)
> > Will answer this question
> 
> @mhfoo Then, if it's okay with you, I suggest to close this issue. Just make sure you review PR #113 to verify that the OIDC profile being defined is aligned with Signtel's expectations and needs.

@jpengar 

This issue will auto close based on the successful merge of the following:
#105 and #113 😀

![image](https://github.com/camaraproject/IdentityAndConsentManagement/assets/134043190/c2f524a6-86cc-4e46-885a-c62239ced879)
